src/coreclr/nativeaot/System.Private.CoreLib/src/System/InvokeUtils.cs

```diff
@@ -128,121 +128,313 @@ private static Exception ConvertOrWidenPrimitivesEnumsAndPointersIfPossible(obje
-                    if (new EETypePtr(srcEEType).CorElementType == CorElementType.ELEMENT_TYPE_CHAR)
+                    switch (srcElementType)
                     {
-                        dstObject = (double)(char)srcObject;
-                    }
-                    else
-                    {
-                        dstObject = Convert.ToDouble(srcObject);
+                        case EETypeElementType.Char:
+                            doubleValue = (double)Unsafe.As<byte, char>(ref rawSrcValue);
+                            break;
+                        case EETypeElementType.SByte:
+                            doubleValue = (double)Unsafe.As<byte, sbyte>(ref rawSrcValue);
+                            break;
+                        case EETypeElementType.Byte:
+                            doubleValue = (double)rawSrcValue;
+                            break;
+                        case EETypeElementType.Int16:
+                            doubleValue = (double)Unsafe.As<byte, short>(ref rawSrcValue);
+                            break;
+                        case EETypeElementType.UInt16:
+                            doubleValue = (double)Unsafe.As<byte, ushort>(ref rawSrcValue);
+                            break;
+                        case EETypeElementType.Int32:
+                            doubleValue = (double)Unsafe.As<byte, int>(ref rawSrcValue);
+                            break;
+                        case EETypeElementType.UInt32:
+                            doubleValue = (double)Unsafe.As<byte, uint>(ref rawSrcValue);
+                            break;
+                        case EETypeElementType.Int64:
+                            doubleValue = (double)Unsafe.As<byte, long>(ref rawSrcValue);
+                            break;
+                        case EETypeElementType.UInt64:
+                            doubleValue = (double)Unsafe.As<byte, ulong>(ref rawSrcValue);
+                            break;
+                        case EETypeElementType.Single:
+                            doubleValue = (double)Unsafe.As<byte, float>(ref rawSrcValue);
+                            break;
+                        case EETypeElementType.Double:
+                            doubleValue = Unsafe.As<byte, double>(ref rawSrcValue);
+                            break;
+                        default:
+                            goto Failure;
                     }
+                    rawDstValue = &doubleValue;
                     break;
 
                 default:
-                    Debug.Fail("Unexpected CorElementType: " + dstElementType + ": Not a valid widening target.");
-                    dstObject = null;
-                    return CreateChangeTypeException(srcEEType, dstEEType, semantics);
+                    goto Failure;
             }
 
+            dstObject = RuntimeImports.RhBox(dstEEType, ref *(byte*)rawDstValue);
             Debug.Assert(dstObject.GetMethodTable() == dstEEType);
```

Comment:
```suggestion
```
Not valuable anymore. (It was valuable before this change when the code required the stars to align to get the right type here.)